![Logo](https://github.com/bcgsc/pepkseq/blob/main/pepkseq-logo.png)

# pepkseq
## De novo [peptide] coding potential calculator using a k-mer approach
### 2023


## Contents

1. [Description](#description)
2. [Implementation and requirements](#implementation)
3. [Install](#install)
4. [Dependencies](#dep)
5. [Documentation](#docs)
6. [Citing pepkseq](#cite)
7. [Credits](#credits)
8. [Running pepkseq](#run)
9. [Test data](#data)
10. [Output](#output)
11. [Algorithm](#algorithm)
12. [Quick reference](#quickref)
13. [Generating plots](#bplot)
14. [License](#license)


## Description <a name=description></a>

pepkseq systematically processes the k-mers of a supplied DNA sequence and computes regions with coding potential, using target codon translation tables

pepkseq is used to quickly survey regions with [protein] coding potential on all 6 frames


## Implementation and requirements <a name=implementation></a>

pepkseq is developed in PERL and runs on any system where PERL is installed.


## Install <a name=install></a>

Clone and enter the pepkseq directory.
<pre>
git clone https://github.com/bcgsc/pepkseq
cd pepkseq
</pre>



## Dependencies <a name=dep></a>

If PERL is installed on your system, you're good to go (no additional libraries needed, nor dependencies).


## Documentation <a name=docs></a>

Refer to the README.md file on how to install and run pepkseq. Read the 


## Citing pepkseq <a name=cite></a>

Thank you for your [![Stars](https://img.shields.io/github/stars/bcgsc/pepkseq.svg)](https://github.com/bcgsc/pepkseq/stargazers) and for using, developing and promoting this free software!

If you use pepkseq in your research, please cite: 


TBD

## Credits <a name=credits></a>
<pre>
pepkseq concept, algorithm design, implementation: Rene L Warren
</pre>


## Running pepkseq <a name=run></a>

<pre>
Usage: ./pepkseq.pl VERSION
 -f FASTA (DNA/RNA) file (required)
 -k length (option, default: -k 90)
 -c genetic code translation table id (option, default: -c 1 [standard])
 -s min. reference FASTA region [size] (bp) to output (option, default: -s 270 bp)
 -v output tsv file (option, -v 1==yes -v 0==no [default])
</pre>

Notes:
<pre>


 -f FASTA (DNA/RNA) file (required)

 -k length (option, default: -k 90)

 -c genetic code translation table id (option, default: -c 1 [standard])

 -s min. reference FASTA region [size] (bp) to output (option, default: -s 270 bp)

 -v output tsv file (option, -v 1==yes -v 0==no [default])


</pre>

### Test data <a name=data></a>
---------

TBD

## Output  <a name=output></a>

1) TSV file 


2) FASTA file (*DNA.fa)

   A multi-FASTA 

3) FASTA file (*PROTEIN.fa)


4) LOG file (.log)

   Captures the verbose STDOUT in a log file, showing the progress of the program.    


## Algorithm design and implementation <a name=algorithm></a>

TBD
 
## Quick reference <a name=quickref></a>

TBD

## Generating plots <a name=bplot></a>

This is an example plot, generated from the coding potential tsv file (-v 1) generated by pepkseq.
![pepkseqPlot](https://github.com/bcgsc/pepkseq/blob/main/CEMAcpPlot.png)
This example predicted coding potential regions is depicted using -k 150 on C. maximus (whale shark) mitogenome with the following command:
<pre>
./pepkseq.pl -f CEMA.fa -k 150 -s 200 -c 2 -v 1
</pre>

R code to generate similar plots
<pre>
##coding potential plots, with R (example provided)

library(ggplot2)
library(readr)
library(scales)
library(ggdark)

df <- read.table('pepkseq_v0.0.1-f_CEMA.fa-k150-frameKmers.tsv', sep="\t", header = TRUE)

ggplot(df, aes(y=coding, x=position, fill = factor(frame))) +
  geom_col() +
  scale_fill_manual(values = c("#b35806","#f1a340","#fee0b6","#d8daeb","#998ec3","#542788")) +
  xlab("White shark mitogenome (ch-CACA1-CM030070.1)") +ylim(-1,1)+ ylab("Coding potential") + dark_theme_gray(base_size = 14)
</pre>

## License <a name=license></a>

pepkseq Copyright (c) 2020-2023 British Columbia Cancer Agency Branch.  All rights reserved.

pepkseq is released under the GNU General Public License v3

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, version 3.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.

For commercial licensing options, please contact
Patrick Rebstein <prebstein@bccancer.bc.ca>
